### PR TITLE
[SPARK-16401] [SQL] Data Source API: Enable Extending RelationProvider and CreatableRelationProvider without Extending SchemaRelationProvider

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

When users try to implement a data source API with extending only `RelationProvider` and `CreatableRelationProvider`, they will hit an error when resolving the relation.

``` Scala
spark.read
.format("org.apache.spark.sql.test.DefaultSourceWithoutUserSpecifiedSchema")
  .load()
  .write.
format("org.apache.spark.sql.test.DefaultSourceWithoutUserSpecifiedSchema")
  .save()
```

The error they hit is like

```
org.apache.spark.sql.test.DefaultSourceWithoutUserSpecifiedSchema does not allow user-specified schemas.;
org.apache.spark.sql.AnalysisException: org.apache.spark.sql.test.DefaultSourceWithoutUserSpecifiedSchema does not allow user-specified schemas.;
    at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:319)
    at org.apache.spark.sql.execution.datasources.DataSource.write(DataSource.scala:494)
    at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:211)
```

Actually, the bug fix is simple. [`DataSource.createRelation(sparkSession.sqlContext, mode, options, data)`](https://github.com/gatorsmile/spark/blob/dd644f8117e889cebd6caca58702a7c7e3d88bef/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L429) already returns a BaseRelation. We should not assign schema to `userSpecifiedSchema`. That schema assignment only makes sense for the data sources that extend `FileFormat`.
#### How was this patch tested?

Added a test case.

### CODE REVIEW

**Comment by gatorsmile** [comment]
@rxin @marmbrus @yhuai Not sure whether we should backport to 2.0. Personally, this sounds a critical bug to me.

**Comment by rxin** [comment]
Is this a regression?

**Comment by gatorsmile** [comment]
Will make a try tonight. My environment for 1.6 is at home.

**Comment by JustinPihony** [comment]
@rxin It does look like this might have been a regression introduced via [the initial creation of `DataSource`](https://github.com/apache/spark/blob/1e28840594b9d972c96d3922ca0bf0f76e313e82/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala) from [`ResolvedDataSource`](https://github.com/apache/spark/blob/branch-1.6/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ResolvedDataSource.scala). Although, @marmbrus could speak to it better as he was the one who wrote that code. Maybe there was a reason?

**Comment by gatorsmile** [comment]
@rxin This is a regression. I did try it in Spark 1.6. It works well. I think we need to fix it in Spark 2.0

Thanks!

**Comment by rxin** [comment]
Thanks - then we should merge this in 2.0.

**Comment by gatorsmile** [comment]
Sure, let me know whether I need to submit another PR for backporting to 2.0. Thanks!

**Comment by rxin** [comment]
cc @cloud-fan and @liancheng for review.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 1]
```diff
@@ -82,6 +82,34 @@ class DefaultSource
...
+
+  case class FakeRelation(sqlContext: SQLContext) extends BaseRelation {
+    override def schema: StructType = StructType(Seq(StructField("a", StringType)))
+  }
+
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String]): BaseRelation = {
+    LastOptions.parameters = parameters
+    LastOptions.schema = None
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Why assign values here? We don't check them in the test right?

**Comment by cloud-fan** [comment]
Nice catch! The fix LGTM.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: That is from the another default source. It is useless in this test case. Let me clean them. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 2]
```diff
@@ -82,6 +82,32 @@ class DefaultSource
...
+  with CreatableRelationProvider {
+
+  case class FakeRelation(sqlContext: SQLContext) extends BaseRelation {
+    override def schema: StructType = StructType(Seq(StructField("a", StringType)))
+  }
+
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String]): BaseRelation = {
+    LastOptions.parameters = parameters
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: hmmm, is it a convention to assign parameters to `LastOptions`? I think the test can work without it.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Yeah, sure. Let me remove all these things. Thanks!

**Comment by cloud-fan** [comment]
thanks, merging to master and 2.0!

**Pull request closed without merging**